### ISSUE TITLE
[TS] Context not required on MachineConfig

### ISSUE THREAD

**Issue by Andarist** [statelyai/xstate#583]
## Description:

Current typings allowed context to stay `undefined` and this might cause a runtime error - https://codesandbox.io/s/xstate-typescript-template-452k1 .

Because a MachineConfig is kinda a static template of the machine which we can later extend with `.withContext` etc this probably needs to stay optional.

I'm wondering what we can do to overcome this - would splitting the type to `MachineConfig` & `MachineConfigWithContext` help here? Assuming ofc that we would, based on this, create 2 distinct return types and make `interpret` accept only the `MachineWithContext`? 

What about "pure" Machine API? I think it might be vulnerable right now in a similar fashion.

**Comment by davidkpiano** [comment]
I think that we should make `config.context` required if `TContext` is not `undefined`. Otherwise I don't think there is any guarantee that a developer might (accidentally) use a machine defined without context directly - and we want to prevent this from happening at compile-time.

I also think we should enforce the following patterns: 

1. **Creating a machine based on context**

```js
const createUserMachine = (name: string) => Machine<{ name: string }>({
  id: 'user',
  context: { name },
  // ...
});
```

2. **Context with explicit optional properties**

```js
const userMachine = Machine<{ name?: string }>({
  id: 'user',
  context: { name: undefined },
  // ...
});
```

**Comment by Andarist** [comment]
Seems to me that even support for `undefined` there brings some unnecessary complexity. IMHO would be better to just default to an empty object there - this would make both implementation and typings easier.

Either way - gonna work on a PR to improve the story around this soon 🚀 

> I also think we should enforce the following patterns:

Agreed.

**Comment by davidkpiano** [comment]
@Andarist Will this be part of the upcoming type generic refactors?

We can check if `TTypes['context'] extends MachineContext` and fail if `TConfig['context'] extends undefined` or something like that

**Comment by davidkpiano** [comment]
Fixed in #4049
